pkg/util/quotapool/intpool.go

```diff
@@ -0,0 +1,236 @@
+// allows the client to provide a function which will be used to determine
+// whether a quantity of quota is sufficient when it becomes available.
+type IntPool struct {
+	qp  *QuotaPool
+	max int64
+
+	intAllocSyncPool       sync.Pool
+	intRequestSyncPool     sync.Pool
+	intFuncRequestSyncPool sync.Pool
+}
+
+// IntAlloc is an allocated quantity which should be released.
+type IntAlloc struct {
+	alloc int64
+	p     *IntPool
+}
+
+// Release releases an IntAlloc back into the IntPool.
+func (ia *IntAlloc) Release() {
+	ia.p.qp.Add((*intAlloc)(ia))
+}
+
+// Acquired returns the quantity that this alloc has acquired.
+func (ia *IntAlloc) Acquired() int64 {
+	return ia.alloc
+}
+
+// Merge adds the acquired resources in other to ia. Other may not be used after
+// it has been merged. It is illegal to merge allocs from different pools and
+// doing so will result in a panic.
+func (ia *IntAlloc) Merge(other *IntAlloc) {
+	if ia.p != other.p {
+		panic("cannot merge IntAllocs from two different pools")
+	}
+	ia.alloc = min(ia.p.max, ia.alloc+other.alloc)
+	ia.p.putIntAlloc(other)
+}
+
+// String formats an IntAlloc as a string.
+func (ia *IntAlloc) String() string {
+	if ia == nil {
+		return strconv.Itoa(0)
+	}
+	return strconv.FormatInt(ia.alloc, 10)
+}
+
+// intAlloc is used to make IntAlloc implement Resource without muddling its
+// exported interface.
+type intAlloc IntAlloc
+
+// Merge makes intAlloc a Resource.
+func (ia *intAlloc) Merge(other Resource) {
+	(*IntAlloc)(ia).Merge((*IntAlloc)(other.(*intAlloc)))
+}
+
+// NewIntPool creates a new named IntPool with a maximum quota value.
+func NewIntPool(name string, max int64, options ...Option) *IntPool {
+	p := IntPool{
+		max: max,
+		intAllocSyncPool: sync.Pool{
+			New: func() interface{} { return new(IntAlloc) },
+		},
+		intRequestSyncPool: sync.Pool{
+			New: func() interface{} { return new(intRequest) },
+		},
+		intFuncRequestSyncPool: sync.Pool{
+			New: func() interface{} { return new(intFuncRequest) },
+		},
+	}
+	p.qp = New(name, (*intAlloc)(p.newIntAlloc(max)), options...)
+	return &p
+}
+
+// Acquire acquires the specified amount of quota from the pool. On success,
+// nil is returned and the caller must call add(v) or otherwise arrange for the
+// quota to be returned to the pool. If 'v' is greater than the total capacity
+// of the pool, we instead try to acquire quota equal to the maximum capacity.
+//
+// Acquisitions of 0 return immediately with no error, even if the IntPool is
+// closed.
+//
+// Safe for concurrent use.
+func (p *IntPool) Acquire(ctx context.Context, v int64) (*IntAlloc, error) {
+	// Special case acquisitions of size 0.
+	if v == 0 {
+		return p.newIntAlloc(v), nil
+	}
+	r := p.newIntRequest(v)
+	defer p.putIntRequest(r)
+	if err := p.qp.Acquire(ctx, r); err != nil {
+		return nil, err
+	}
+	return p.newIntAlloc(v), nil
+}
+
+// IntRequestFunc is used to request a quantity of quota determined when quota is
+// available rather than before requesting.
```

Comment:
When is this useful? And when is it called? Is it called every time quota is returned to the pool? I'm struggling to come up with the obvious example here. Maybe add one?

```go
// For example, an application wanting to ... would use an IntRequestFunc as below:
// func(....) (...) { }
```